go/plugins/compat_oai/openai/generate.go

```diff
@@ -0,0 +1,132 @@
+		case ai.RoleTool:
+			oaiMessages = append(oaiMessages, openai.ToolMessage("", content)) // TODO: Add tool ID if needed
+		default:
+			oaiMessages = append(oaiMessages, openai.UserMessage(content))
+		}
+	}
+	g.request.Messages = openai.F(oaiMessages)
+	return g
+}
+
+// WithConfig adds configuration parameters from the model request
+func (g *ModelGenerator) WithConfig(config any) *ModelGenerator {
+	if config != nil {
+		// TODO: Implement configuration from model request
+		// modelRequest.Config is any type
+	}
+	return g
+}
+
+// WithTools adds tools to the request
+func (g *ModelGenerator) WithTools(tools []ai.Tool, choice ai.ToolChoice) *ModelGenerator {
+	// TODO: Implement tools from model request
+	// see vertex ai recent pr here for reference: https://github.com/firebase/genkit/pull/2259
+	return g
+}
+
+// Generate executes the generation request
+func (g *ModelGenerator) Generate(ctx context.Context, handleChunk func(context.Context, *ai.ModelResponseChunk) error) (*ai.ModelResponse, error) {
+	if handleChunk != nil {
+		return g.generateStream(ctx, handleChunk)
+	}
+	return g.generateComplete(ctx)
+}
+
+// concatenateContent concatenates text content into a single string
+func (g *ModelGenerator) concatenateContent(parts []*ai.Part) string {
+	content := ""
+	for _, part := range parts {
+		content += part.Text
+	}
+	return content
+}
+
+// generateStream generates a streaming model response
+func (g *ModelGenerator) generateStream(ctx context.Context, handleChunk func(context.Context, *ai.ModelResponseChunk) error) (*ai.ModelResponse, error) {
+	stream := g.client.Chat.Completions.NewStreaming(ctx, *g.request)
+	defer stream.Close()
+
+	var fullResponse ai.ModelResponse
+	fullResponse.Message = &ai.Message{
+		Role:    ai.RoleModel,
+		Content: make([]*ai.Part, 0),
+	}
+
+	for stream.Next() {
+		chunk := stream.Current()
+		if len(chunk.Choices) > 0 {
+			content := chunk.Choices[0].Delta.Content
+			modelChunk := &ai.ModelResponseChunk{
+				Content: []*ai.Part{ai.NewTextPart(content)},
+			}
+
+			if err := handleChunk(ctx, modelChunk); err != nil {
+				return nil, fmt.Errorf("callback error: %w", err)
+			}
+
+			fullResponse.Message.Content = append(fullResponse.Message.Content, modelChunk.Content...)
+		}
+	}
+
+	if err := stream.Err(); err != nil {
+		return nil, fmt.Errorf("stream error: %w", err)
+	}
+
+	return &fullResponse, nil
+}
+
+// generateComplete generates a complete model response
+func (g *ModelGenerator) generateComplete(ctx context.Context) (*ai.ModelResponse, error) {
+	completion, err := g.client.Chat.Completions.New(ctx, *g.request)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create completion: %w", err)
+	}
+
+	return &ai.ModelResponse{
+		Message: &ai.Message{
+			Role: ai.RoleModel,
+			Content: []*ai.Part{
+				ai.NewTextPart(completion.Choices[0].Message.Content),
+			},
+		},
+		FinishReason: ai.FinishReason("stop"),
```

Comment:
Without this change, the test fails.

```suggestion
		FinishReason: ai.FinishReason("stop"),
		Request:      &ai.ModelRequest{},
		Usage: &ai.GenerationUsage{
			InputTokens:  int(completion.Usage.PromptTokens),
			OutputTokens: int(completion.Usage.CompletionTokens),
			TotalTokens:  int(completion.Usage.TotalTokens),
		},
```

Reply:
thanks - added!